File: bindings/go/statepersistence/statepersistence.go
package statepersistence

import (
    "encoding/json"
    "os"
)

type Aggregator struct {
    SocketPath       string  `json:"socket_path"`
    SocketPermission *uint32 `json:"socket_permission,omitempty"`
}

type GitConfig struct {
    DefaultServer   string `json:"default_server"`
    CredentialsFile string `json:"credentials_file"`
}

type DatabaseConfig struct {
    URL      string `json:"url"`
    PoolSize uint32 `json:"pool_size"`
}

type AppConfig struct {
    AppName     string          `json:"app_name"`
    MaxRamUsage uint64          `json:"max_ram_usage"`
    MaxCpuUsage uint64          `json:"max_cpu_usage"`
    Environment string          `json:"environment"`
    DebugMode   bool            `json:"debug_mode"`
    LogLevel    string          `json:"log_level"`
    Git         *GitConfig      `json:"git,omitempty"`
    Database    *DatabaseConfig `json:"database,omitempty"`
    Aggregator  *Aggregator     `json:"aggregator,omitempty"`
}

type ErrorItem struct {
    ErrType string `json:"err_type"`
    ErrMesg string `json:"err_mesg"`
}

type Output struct {
    Timestamp uint64 `json:"timestamp"`
    Line      string `json:"line"`
}

type AppState struct {
    Name             string      `json:"name"`
    Version          string      `json:"version"`
    Data             string      `json:"data"`
    Status           string      `json:"status"`
    PID              uint32      `json:"pid"`
    LastUpdated      uint64      `json:"last_updated"`
    StaredAt         uint64      `json:"stared_at"`
    EventCounter     uint32      `json:"event_counter"`
    ErrorLog         []ErrorItem `json:"error_log"`
    Config           AppConfig   `json:"config"`
    SystemApplication bool       `json:"system_application"`
    Stdout           []Output    `json:"stdout"`
    Stderr           []Output    `json:"stderr"`
}

func SaveState(path string, state *AppState) error {
    data, err := json.MarshalIndent(state, "", "  ")
    if err != nil {
        return err
    }
    return os.WriteFile(path, data, 0o600)
}

func LoadState(path string) (*AppState, error) {
    data, err := os.ReadFile(path)
    if err != nil {
        return nil, err
    }
    var state AppState
    if err := json.Unmarshal(data, &state); err != nil {
        return nil, err
    }
    return &state, nil
}

File: bindings/python/state_persistence.py
from dataclasses import dataclass, field, asdict
from typing import Optional, List, Tuple
import json


@dataclass
class Aggregator:
    socket_path: str
    socket_permission: Optional[int] = None


@dataclass
class GitConfig:
    default_server: str
    credentials_file: str


@dataclass
class DatabaseConfig:
    url: str
    pool_size: int


@dataclass
class AppConfig:
    app_name: str
    max_ram_usage: int
    max_cpu_usage: int
    environment: str
    debug_mode: bool
    log_level: str
    git: Optional[GitConfig] = None
    database: Optional[DatabaseConfig] = None
    aggregator: Optional[Aggregator] = None


@dataclass
class ErrorItem:
    err_type: str
    err_mesg: str


@dataclass
class AppState:
    name: str
    version: str
    data: str
    status: str
    pid: int
    last_updated: int
    stared_at: int
    event_counter: int
    error_log: List[ErrorItem] = field(default_factory=list)
    config: AppConfig = field(default_factory=AppConfig)
    system_application: bool = False
    stdout: List[Tuple[int, str]] = field(default_factory=list)
    stderr: List[Tuple[int, str]] = field(default_factory=list)


class StatePersistence:
    """Simple JSON-based state persistence."""

    @staticmethod
    def save_state(state: AppState, path: str) -> None:
        with open(path, "w", encoding="utf-8") as fh:
            json.dump(asdict(state), fh, indent=2)

    @staticmethod
    def load_state(path: str) -> AppState:
        with open(path, "r", encoding="utf-8") as fh:
            data = json.load(fh)
        data["config"] = AppConfig(**data["config"])
        data["error_log"] = [ErrorItem(**e) for e in data.get("error_log", [])]
        return AppState(**data)
